src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs

```diff
@@ -0,0 +1,483 @@
+                _onOpen = null;
+            }
+            if (_onMessage != null)
+            {
+                _innerWebSocket?.SetObjectProperty("onmessage", "");
+                _onMessage = null;
+            }
+        }
+
+        public override void Dispose()
+        {
+            System.Diagnostics.Debug.WriteLine("BrowserWebSocket::Dispose");
+            int priorState = Interlocked.Exchange(ref _state, (int)InternalState.Disposed);
+            if (priorState == (int)InternalState.Disposed)
+            {
+                // No cleanup required.
+                return;
+            }
+
+            // registered by the CancellationTokenSource cts in the connect method
+            _cts.Cancel(false);
+            _cts.Dispose();
+
+            _writeBuffer?.Dispose();
+            _receiveMessageQueue.Writer.Complete();
+
+            NativeCleanup();
+
+            _innerWebSocket?.Dispose();
+        }
+
+        // This method is registered by the CancellationTokenSource cts in the connect method
+        // and called by Dispose or Abort so that any open websocket connection can be closed.
+        private async void AbortRequest()
+        {
+            if (State == WebSocketState.Open)
+            {
+                await CloseAsyncCore(WebSocketCloseStatus.NormalClosure, SR.net_WebSockets_Connection_Aborted, CancellationToken.None).ConfigureAwait(continueOnCapturedContext: true);
+            }
+        }
+
+        /// <summary>
+        /// Send data on <see cref="System.Net.WebSockets.ClientWebSocket"/> as an asynchronous operation.
+        /// </summary>
+        /// <returns>The async.</returns>
+        /// <param name="buffer">Buffer.</param>
+        /// <param name="messageType">Message type.</param>
+        /// <param name="endOfMessage">If set to <c>true</c> end of message.</param>
+        /// <param name="cancellationToken">Cancellation token.</param>
+        public override Task SendAsync(ArraySegment<byte> buffer, WebSocketMessageType messageType, bool endOfMessage, CancellationToken cancellationToken)
+        {
+            ThrowIfNotConnected();
+
+            if (messageType != WebSocketMessageType.Binary &&
+                    messageType != WebSocketMessageType.Text)
+            {
+                throw new ArgumentException(SR.Format(SR.net_WebSockets_Argument_InvalidMessageType,
+                    messageType,
+                    nameof(SendAsync),
+                    WebSocketMessageType.Binary,
+                    WebSocketMessageType.Text,
+                    nameof(CloseOutputAsync)),
+                    nameof(messageType));
+            }
+
+            WebSocketValidate.ValidateArraySegment(buffer, nameof(buffer));
+
+            _writeBuffer ??= new MemoryStream();
+            _writeBuffer.Write(buffer.Array!, buffer.Offset, buffer.Count);
+
+            if (!endOfMessage)
+                return Task.CompletedTask;
```

Comment:
The browser websocket API isn't capable of sending messages split over multiple fragments?  The `endOfMessage` argument elsewhere translates into the FIN bit in the websocket frame.

Reply:
We do not have access to the websocket frame through the browser api.  Also, from my understanding there is not the concept of EndOfMessage.  Not that I can find anyway.

Reply:
https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/send expects the full message so we need to buffer somewhere.  An alternative to the MemoryStream buffer would be constructing a chain of Blobs on the JS side and calling send on that when endOfMessage is true.

Reply:
It's fine for the buffering to be in managed.  I was questioning the buffering all-up, but if the websocket API doesn't support fragments, then ok.